common/djangoapps/third_party_auth/middleware.py

```diff
@@ -23,3 +25,46 @@ def get_redirect_uri(self, request, exception):
             redirect_uri = pipeline.AUTH_DISPATCH_URLS[auth_entry]
 
         return redirect_uri
+
+
+class ResetSessionIfPipelineBrokenMiddleware(object):
+    """
+    Middleware signs the user out if they need to provide data sharing consent,
+    haven't, and left the TPA pipeline prematurely
+    """
+    def process_view(self, request, view_func, view_args, view_kwargs):
+        """
+        Conditionally sign out users who don't provide data sharing consent
+        """
+        allowed_module_prefixes = (
```

Comment:
@haikuginger suggestion: it looks like `allowed_module_prefixes` should not "block" any method on the TPA pipeline. With current implementation you use the fact that pipeline only contains methods from `third_party_auth.` and `social.`, but we already have some use cases to extend the pipeline dynamically. So, I would suggest checking method against pipeline first.

Also, there's [AUTH_CUSTOM_FORMS](https://github.com/edx/edx-platform/blob/master/common/djangoapps/third_party_auth/pipeline.py#L120) feature that allows custom form to be included into the pipeline - it might cause some trouble too. As far as I can tell, it is not used in "normal" edX instances, but at least one "solutions" instance uses it in production (codename Apros :)).